flink-runtime/src/test/java/org/apache/flink/runtime/shuffle/ShuffleMasterTest.java

```diff
@@ -0,0 +1,239 @@
+    public static class TestShuffleServiceFactory extends NettyShuffleServiceFactory {
+        @Override
+        public NettyShuffleMaster createShuffleMaster(ShuffleMasterContext shuffleMasterContext) {
+            return new TestShuffleMaster(shuffleMasterContext.getConfiguration());
+        }
+    }
+
+    /** An {@link ShuffleMaster} implementation for testing. */
+    public static class TestShuffleMaster extends NettyShuffleMaster {
+
+        private static final AtomicReference<TestShuffleMaster> currentInstance =
+                new AtomicReference<>();
+
+        private static final BlockingQueue<String> partitionEvents = new LinkedBlockingQueue<>();
+
+        private final AtomicBoolean started = new AtomicBoolean();
+
+        private final AtomicBoolean closed = new AtomicBoolean();
+
+        private final BlockingQueue<ResultPartitionID> partitions = new LinkedBlockingQueue<>();
+
+        private final AtomicReference<JobShuffleContext> jobContext = new AtomicReference<>();
+
+        private final boolean stopTrackingPartition;
+
+        public TestShuffleMaster(Configuration conf) {
+            super(conf);
+            this.stopTrackingPartition = conf.getBoolean(STOP_TRACKING_PARTITION_KEY, false);
+            currentInstance.set(this);
+        }
+
+        @Override
+        public void start() throws Exception {
+            assertFalse(started.get());
+            assertFalse(closed.get());
+            started.set(true);
+            super.start();
+        }
+
+        @Override
+        public void close() throws Exception {
+            assertShuffleMasterAlive();
+            closed.set(true);
+            super.close();
+        }
+
+        @Override
+        public void registerJob(JobShuffleContext context) {
+            assertShuffleMasterAlive();
+            assertTrue(jobContext.compareAndSet(null, context));
+            super.registerJob(context);
+        }
+
+        @Override
+        public void unregisterJob(JobID jobID) {
+            assertJobRegistered();
+            jobContext.set(null);
+            super.unregisterJob(jobID);
+        }
+
+        @Override
+        public CompletableFuture<NettyShuffleDescriptor> registerPartitionWithProducer(
+                JobID jobID,
+                PartitionDescriptor partitionDescriptor,
+                ProducerDescriptor producerDescriptor) {
+            assertJobRegistered();
+            partitionEvents.add("registerPartitionWithProducer");
+
+            CompletableFuture<NettyShuffleDescriptor> future = new CompletableFuture<>();
+            try {
+                NettyShuffleDescriptor shuffleDescriptor =
+                        super.registerPartitionWithProducer(
+                                        jobID, partitionDescriptor, producerDescriptor)
+                                .get();
+                if (partitions.size() == 1 && stopTrackingPartition) {
```

Comment:
It's better to add some explanation for what we want to achieve here. I feel it is not easy to understand at the first glance.